### DISCUSSION TITLE
TLDR; how can I use @layer with createTheme?

### DISCUSSION THREAD

**Discussion by olivierpascal** [vanilla-extract-css/vanilla-extract#1472] [Q&A] [answered]
TLDR; how can I use @layer with `createTheme`?

Hello, I am using vanilla extract and sprinkles, and I really appreciate both.

When we use sprinkles, we usually want that sprinkles styles have a higher priority than base styles. So we use CSS layers, which is well supported by `style()` and `defineProperties()`.

Problems start when we use theme vars inside sprinkles. `createTheme()` does not support layer. So theme vars always have a higher priority than any layer.

What I would like is being able to provide a CSS layer to `createTheme()` so that the CSS class is created inside that layer.

Thanks!

**Comment by olivierpascal** [comment]
Hello 🥺

**Reply by askoufis** [reply]
Unfortunately you can't do this with the current `createTheme` API. You can define a _global_ theme inside a layer, though that may not suit your use case:

```ts
import { layer, createGlobalTheme } from "@vanilla-extract/css";

export const app = layer("app");

export const vars = createGlobalTheme(`@layer ${app}`, {
  color: {
    brand: "blue",
  },
  font: {
    body: "arial",
  },
});
```

I'm struggling to think of a nice API to define a theme within a layer without a breaking change. Will have to get back to you on this one.

**Reply by olivierpascal** [reply]
Hugh. Thank you so much

**Reply by olivierpascal** [reply]
Hello @askoufis, by any chance, did you find that nice API yet? 🙏🏼

**Reply by olivierpascal** [reply]
Hello, may I ask if we can use @layer with createTheme() yet or if this is planned soon? This is for me the missing part of the sprinkles. Many thanks 🙏🏼

**Reply by askoufis** [reply]
Sorry for the delayed response. I think a special `@layer` key inside `createTheme` would be the way to go. This is consistent with [how sprinkles are assigned to layers](https://vanilla-extract.style/documentation/packages/sprinkles/#layer) too.

I didn't realize this in my original response to your questio, but you can manually create a theme inside a layer using a combination of `style`, `layer`, `createThemeContract` and `assignVars`:

```ts
import {
  style,
  createThemeContract,
  assignVars,
  layer,
} from "@vanilla-extract/css";

export const vars = createThemeContract({
  colors: {
    primary: null,
    secondary: null,
  },
  space: {
    small: null,
    large: null,
  },
});

const themeLayer = layer("theme");

export const themeInLayer = style({
  "@layer": {
    [themeLayer]: {
      vars: assignVars(vars, {
        colors: {
          primary: "red",
          secondary: "blue",
        },
        space: {
          small: "10px",
          large: "20px",
        },
      }),
    },
  },
});
```

**Reply by olivierpascal** [reply]
Hello, thanks for the suggestion. But it add a lot of overhead because all vars has to be redefined.

I agree with a special `@layer` key inside `createTheme`.

Do you think it could be implemented i a next update of vanilla-extract-css?

**Reply by askoufis** [reply]
Sure, the theme needs to be assigned values separately from the contract being created. This isn't any different from how contracts and themes are usually created with `createThemeContract` and `createTheme`. `themeInLayer` is sort of a re-implementation of `createTheme` but using a layer.

I can take a look at adding layer support to `createTheme`. It shouldn't be too complicated to implement, however I don't want to make any promises around a release timeline.

**Reply by olivierpascal** [reply]
That would really be amazing...

**Reply by askoufis** [reply, marked as answer]
@olivierpascal `@layer` support in `createTheme` and `createGlobalTheme` has been released in [`@vanilla-extract/css@1.17.0`](https://github.com/vanilla-extract-css/vanilla-extract/releases/tag/%40vanilla-extract/css%401.17.0).

**Reply by olivierpascal** [reply]
It works like a charm, thank you very much!
